### DISCUSSION TITLE
How can I pass along my `TFieldValues` generic to a wrapped controlled input?

### DISCUSSION THREAD

**Discussion by iKrushYou** [react-hook-form/react-hook-form#5541] [Q&A] [answered]
Hey, while using RHF i have run into this TS issue. now, I am still fairly new to TS and this is a more advanced usage so I'm not sure if it's possible even. And even though this isn't an issue specifically with RHF I would like to ask here in hopes somebody might know how to handle this situation!

https://codesandbox.io/s/react-hook-form-generic-input-wrapper-6kpyv?file=/src/App.tsx

What would be the best way to capture the types from the `control` to pass into a generic input wrapper? specifically the inputs for the `FormInput` lines 13-16

I haven't yet quite mastered generics. I know that I want to capture the `<TFieldValues>` but I'm not sure how to do that with a component (instead of a function)

I know that if it were a function I would just pass along the type like `myFunction<FormProps>` but I'm not sure how to do that with a component

much thanks!

**Comment by bluebill1049** [comment]
the `string` part should be replaced with https://react-hook-form.com/ts#FieldPath

**Reply by iKrushYou** [reply]
Hey, thanks for the response. I knew what types i should be using but the problem i was running into was actually implementing generic into my typescript functional components. I managed to figure it out, though!

https://codesandbox.io/s/react-hook-form-generic-input-wrapper-forked-hwuf5

```jsx
const FormInput = <
  TFieldValues extends FieldValues,
  TName extends FieldPath<TFieldValues>
>({
  name,
  label,
  control,
  defaultValue
}: {
  label?: string;
} & UseControllerProps<TFieldValues, TName>) => {
  return (
    <Controller
      name={name}
      control={control}
      render={({ field }) => (
        <label>
          {label}
          <input {...field} />
        </label>
      )}
      defaultValue={defaultValue}
    />
  );
};
```

**Reply by bluebill1049** [reply]
you want to move this into another comment and set an answer to help others?

**Comment by iKrushYou** [comment, marked as answer]
Solution:

https://codesandbox.io/s/react-hook-form-generic-input-wrapper-forked-hwuf5

```jsx
const FormInput = <
  TFieldValues extends FieldValues,
  TName extends FieldPath<TFieldValues>
>({
  name,
  label,
  control,
  defaultValue
}: {
  label?: string;
} & UseControllerProps<TFieldValues, TName>) => {
  return (
    <Controller
      name={name}
      control={control}
      render={({ field }) => (
        <label>
          {label}
          <input {...field} />
        </label>
      )}
      defaultValue={defaultValue}
    />
  );
};
```

**Reply by minhchu** [reply]
@iKrushYou Hello, I still see a warning regarding `name` attribute: `Type '"namex"' is not assignable to type '"name"'`
https://codesandbox.io/s/react-hook-form-generic-input-wrapper-forked-hwuf5?file=/src/App.tsx:777-845

Do you have any idea to fix that ?

**Reply by iKrushYou** [reply]
> @iKrushYou Hello, I still see a warning regarding `name` attribute: `Type '"namex"' is not assignable to type '"name"'`
> https://codesandbox.io/s/react-hook-form-generic-input-wrapper-forked-hwuf5?file=/src/App.tsx:777-845
> 
> Do you have any idea to fix that ?

Hey I think the CSB got edited during testing. Just rename the namex back to name

**Reply by iKrushYou** [reply]
Okay I fixed the example it should work now

**Reply by vamcs** [reply]
Thanks, this helped me find a bug with my types! 😁 

Some other tips I can add for others:

- The prop type can be simplified with `UseControllerProps<T> & ...`, since `TName` is automatically inferred internally. 
- Also, whenever you send in `control` to the custom component, `T` (or `TFieldValues` in this example) will be automatically inferred too, so you won't have to call the component as `<MyCustomComponent<T> ... />`, so `<MyCustomComponent control={control} ... />` without the generics will be enough!

**Comment by artgladun** [comment]
Could we continue the thread? 

I am trying to implement a reusable section like your address; 
Assume we have a reusable adress section with different name;

```
interface IFormValues {
   name: string;
   invoiceAddress: IAddress;
   destinationAddress: IAddress;
}

```

Is there a chance to have a correct typings for 

```
   <Addresssection name="invoiceAddress" control={control} />
   <Addresssection name="destinationAddress" control={control} />
```

```
// TODO correct typings 
const AddressSection = ({ name, control }) => {

  return (
      <FormInput name={`${name}.street`} control={control} />
      <FormInput name={`${name}.city`} control={control} />
  )
}

````

I've tried using such types as FormFieldValues
```
type NestedUserFormValuesType<T extends string = string, V> = {
  [key in T]: IAddress;
}
````
and put it as a FieldValues; but without success 

what is a good practice to create such reusable form sections?

**Reply by artgladun** [reply]
@bluebill1049 @iKrushYou

**Reply by iKrushYou** [reply]
> @bluebill1049 @iKrushYou

Hey I tried for like 30m to figure out some sort of a solution but unfortunately was not successful with how you have it set up.. The best thing I would recommend is maybe using a field array of addresses and you can make that reusable

**Comment by artgladun** [comment]
This is what I came up with

```

export type AddressFormValues = {
  street: string;
  city: string;
};

interface FormAddressInputProps<TName> {
  name: TName;
}

type FormAddressInputType = <FormValues, TName extends FieldPath<FormValues> = FieldPath<FormValues>>(
  props: FormAddressInputProps<TName>
) => React.ReactElement;

export const FormAddressInput: FormAddressInputType = ({ name }) => {
  const streetName: PathImpl<typeof name, AddressFormValues> = `${name}. street`;
  const cityName: PathImpl<typeof name, AddressFormValues> = `${name}. city`;

  const { control } = useFormContext();

  return (
    <>
       <FormTextInput name={streetName} control={control} />
       <FormTextInput name={cityName} control={control} />
    </>
  );
};

```

```
interface IExampleFormValues {
   invoiceAddress: AddressFormValues;
}

 <FormAddressInput<IExampleFormValues> name="invoiceAddress" />
```
